packages/flutter/lib/src/material/chip.dart

```diff
@@ -1835,7 +1885,14 @@ class _RawChipState extends State<RawChip> with TickerProviderStateMixin<RawChip
     final ThemeData theme = Theme.of(context)!;
     final ChipThemeData chipTheme = ChipTheme.of(context);
     final TextDirection? textDirection = Directionality.of(context);
-    final ShapeBorder shape = widget.shape ?? chipTheme.shape;
+    final BorderSide? resolvedSide = MaterialStateProperty.resolveAs<BorderSide?>(
```

Comment:
I'm afraid this needs to be more complicated. If the widget's BorderSide is a MaterialStateProperty and it resolves to null, then we should defer to the theme's BorderSide, and if _that_ resolves to null, then we should use a default value. Likewise for the shape.

That way an example, like the one you included with this PR's description could just override the default border for the focused state:

```dart
Chip(
  side: _MaterialStateBorderSide((states) {
    if (states.contains(MaterialState.focused)) {
      return const BorderSide(
        color: Colors.red,
        width: 1,
      );
    }
    return null; // defer to the theme or widget default
  }),
)
```

ButtonStyleButton.build() does this kind of MaterialStateProperty resolution.

Also: since we're going to allow this property to be a BorderSide or a MaterialStateProperty, we should add MaterialStateBorderSide and MaterialStateOutlinedShape to material_state.dart.

Reply:
You're right. Thanks for catching this. I'll add a test for it as well.

I'll add the abstract classes `MaterialStateBorderSide` and `MaterialStateOutlinedShape` then, so users can decide for example what border it should be extending (i.e. StadiumBorder or something else).

Reply:
I must also set the new abstract classes to be nullable: `MaterialStateProperty<BorderSide?>`. That differs from the `MaterialStateColor`, and `MaterialStateMouseCursor` which are both non-nullable, i.e. `MaterialStateProperty<MouseCursor>`. Might be confusing to have different kind of styles of APIs here, but I think the benefit of having them nullable is still better, to be able to defer back to default values from the theme or widget.